lib/kino/screen.ex

```diff
@@ -0,0 +1,325 @@
+defmodule Kino.Screen do
+  @moduledoc ~S'''
+  Provides a LiveView like experience for building forms in Livebook.
+
+  Each screen must implement the `c:init/1` and `c:render/1` callbacks.
+  Event handlers can be attached by calling the `control/2` function.
+
+  Each user on the page get their own version of their screen, which
+  is not shared. If you want to share them, you can pass a frame as
+  argument to the screen, as shown in the "Dynamic select" example below.
+
+  ## Dynamic form/select
+
+  Here is an example that allows you to build a dynamic form that renders
+  values depending on the chosen options. On submit, you then process the
+  data (with optinal validation).
+
+      defmodule MyScreen do
+        @behaviour Kino.Screen
+
+        # Import Kino.Control for forms, Kino.Input for inputs, and Screen for control/2
+        import Kino.{Control, Input, Screen}
+
+        @countries [none: "", usa: "United States", canada: "Canada"]
+
+        @languages [
+          usa: [none: "", en: "English", es: "Spanish"],
+          canada: [none: "", en: "English", fr: "French"]
+        ]
+
+        @defaults %{
+          country: :none,
+          language: :none
+        }
+
+        # This is a function we will use to start the screen.
+        #
+        # Our screen will be placed in a grid with one additional
+        # frame to render results into. And the state of the screen
+        # holds the form data and the result frame itself.
+        def new(result_frame) do
+          result_frame = Kino.Frame.new()
+          state = %{data: @defaults, frame: result_frame}
+
+          Kino.Layout.grid([
+            Kino.Screen.new(__MODULE__, state),
+            result_frame
+          ])
+        end
+
+        def render(%{data: data}) do
+          form(
+            [
+              country: country_select(data),
+              language: language_select(data)
+            ]
+            |> Enum.filter(fn {_key, value} -> value != nil end),
+            report_changes: true,
+            submit: "Submit"
+          )
+          |> control(&handle_event/2)
+        end
+
+        defp country_select(data) do
+          select("Country", @countries, default: data.country)
+        end
+
+        defp language_select(data) do
+          if languages = @languages[data.country] do
+            default = if languages[data.language], do: data.language, else: :none
+            select("Language", languages, default: default)
+          end
+        end
+
+        def handle_event(%{data: data, type: :change}, state) do
+          %{state | data: Map.merge(@defaults, data)}
```

Comment:
@jonatanklosko we could remove this merge if we allow a form to have an input set to nil. In that case, the input is not rendered by the data comes back set as nil. Then I can change the default value to `nil: ""` and streamline this code a bit more.

Reply:
I see how that would be convenient, but it does feel weird outside of this context. My main concern though, is that it would crash older Livebook versions. Perhaps we could do some special casing and transform the event before broadcasting in Kino, but I am not a fan of that either :<

Reply:
@jonatanklosko I think it could be used in other contexts, basically every time you need to build a form dynamically and some fields may be missing, we need to do pruning and then, when accessing the `data`, use `Map.get(data, field)` for those fields.